thorlcr/thorutil/thbuf.cpp

```diff
@@ -606,6 +608,549 @@ class CSmartRowInMemoryBuffer: public CSimpleInterface, implements ISmartRowBuff
     }
 };
 
+//#define TRACE_SPILLING_ROWSTREAM // traces each row read/written, and other events
+
+// based on query that produces records with a single sequential (from 1) unsigned4
+// #define VERIFY_ROW_IDS_SPILLING_ROWSTREAM
+
+#define STRESSTEST_SPILLING_ROWSTREAM // 'stressLookAhead' code, which when enabled, reduces buffer sizes etc. to stress test the lookahead spilling
+
+
+
+/* CCompressedSpillingRowStream implementation details:
+ - Writer:
+ - The writer to an in-memory queue, and when the queue is full, or a certain number of rows have been queued, it writes to starts writing to temp files.
+ - The writer will always write to the queue if it can, even after it has started spilling.
+ - The writer commits to disk at LookAheadOptions::writeAheadSize granularity
+ - The writer creates a new temp file when the current one reaches LookAheadOptions::tempFileGranularity
+ - The writer pushes the current nextOutputRow to a queue when it creates the next output file (used by the reader to know when to move to next)
+ - NB: writer implements ISmartRowBuffer::flush() which has slightly weird semantics (blocks until everything is read or stopped)
+-  Reader:
+ - The reader will read from the queue until it is exhausted, and block to be signalled for more.
+ - If the reader dequeues a row that is ahead of the expected 'nextInputRow', it will stash it, and read from disk until it catches up to that row.
+ - If the reader is reading from disk and it catches up with 'committedRows' it will block until the writer has committed more rows.
+ - When reading from a temp file, it will take ownership the CFileOwner and dispose of the underlying file when it has consumed it.
+ - The reader will read from the stream until it hits 'currentTempFileEndRow' (initially 0), at which point it will open the next temp file.
+ */
+
+// NB: Supports being read by 1 thread and written to by another only
+class CCompressedSpillingRowStream: public CSimpleInterfaceOf<ISmartRowBuffer>, implements IRowWriter
+{
+    typedef std::tuple<const void *, rowcount_t, size32_t> RowEntry;
+
+    CActivityBase &activity; // ctor input parameter
+    StringAttr baseTmpFilename; // ctor input parameter
+    LookAheadOptions options; // ctor input parameter
+    Linked<ICompressHandler> compressHandler; // ctor input parameter
+
+    // derived from input paramter (IThorRowInterfaces *rowIf)
+    Linked<IOutputMetaData> meta;
+    Linked<IOutputRowSerializer> serializer;
+    Linked<IEngineRowAllocator> allocator;
+    Linked<IOutputRowDeserializer> deserializer;
+    memsize_t compressionBlockSize = 0; // filled in createOutputStream
```

Comment:
move into the options structure?  It would keep them together, and may make it easier to common up the code with the splitter.